src/main/java/com/jakub/todoSandbox/repository/InMemoryRepository.java

```diff
@@ -0,0 +1,165 @@
+package com.jakub.todoSandbox.repository;
+
+import com.jakub.todoSandbox.model.Step;
+import com.jakub.todoSandbox.model.Todo;
+import org.springframework.stereotype.Repository;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Repository
+public class InMemoryRepository implements TodoRepository{
+
+    private final Map<Long, Todo> todoList = new HashMap<Long, Todo>();
+    private static final int MAX_NUM_STEPS = 10;
+    private Long maxTodoId = 0L;
+
+    @Override
+    public Todo findTodoById(Long todoId) {
+        for (Todo todo : todoList.values()) {
+            if (todo.id().equals(todoId)) {
+                return todoList.get(todoId);
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public List<Todo> findAllTodos() {
+        List<Todo> todoMapToList = new ArrayList<>(todoList.values());
+        return sortByPriority(todoMapToList);
+    }
+
+    @Override
+    public Todo saveTodo(Todo todo) {
+        if (validateNameAndDesc(todo.name(), todo.description())) {
+            var toBeSavedTodo = new Todo((maxTodoId + 1L), todo.name(), todo.description(), todo.priority(), todo.steps());
```

Comment:
Two things:
1. Validation like that should be above, in the service layer
2. Why we return null, if validation was not succesfull? We should throw some custom exception

Reply:
@Iprogrammerr  is the service layer really required just to contain the validation methods?

Reply:
Mostly it is put there. In some cases, you can put this in the Controller, but you shouldn't have it in the repository. Repository is only to store data, nothing else.